arangod/Graph/Enumerators/YenEnumerator.h

```diff
@@ -0,0 +1,275 @@
+    // allocation, since it only consists of two uint64_t values. This is
+    // why we only act on coordinators here:
+    if (!ServerState::instance()->isCoordinator()) {
+      return item;  // A copy, but this is cheap!
+    }
+    uint8_t const* p = item.vpack();
+    VPackSlice v{p};
+    velocypack::ValueLength s = v.byteSize();
+    auto q = (uint8_t const*)(makeLocalCopy((char const*)p, (size_t)s));
+    return EdgeDocumentToken(VPackSlice{q});
+  }
+
+ private:
+  char* makeLocalCopy(char const* p, size_t s) {
+    auto* batch = &_data.back();
+    if (s > batch->buffer.size() - batch->nextFree) {
+      if (s <= BATCH_SIZE) {
+        _data.emplace_back();
+        _totalSize += BATCH_SIZE;
+        _resourceMonitor.increaseMemoryUsage(BATCH_SIZE);
+      } else {
+        _data.emplace_back(s);
+        _totalSize += s;
+        _resourceMonitor.increaseMemoryUsage(s);
+      }
+      batch = &_data.back();
+    }
+    char* place = &batch->buffer[batch->nextFree];
+    batch->nextFree += s;
+    memcpy(place, p, s);
+    return place;
+  }
+};
+
+class PathValidatorOptions;
+struct TwoSidedEnumeratorOptions;
+
+template<class ProviderType, class Step>
+class PathResult;
+
+// This is a relatively straightforward implementation of Yen's algorithm
+// as described here: https://en.wikipedia.org/wiki/Yen%27s_algorithm
+// It uses single shortest path algorithms by means of a subobject of
+// type `TwoSidedEnumerator` (in the unweighted case) and
+// `WeightedTwoSidedEnumerator` (in the weighted case).
+// The subtlety lies in the template magic going on all around this.
+// The `ProviderType` is used to get actual graph data like neighbours
+// out of the system. Essentially, there can be a `SingleServerProvider`
+// or a `ClusterProvider`, and it can be wrapped with a tracer wrapper or
+// not. This provider is just handed on to the `EnumeratorType`, which is
+// essentially an instance of the `TwoSidedEnumerator` or the
+// `WeightedTwoSidedEnumerator`. However, we have to be able to forbid
+// some vertices and edges. This is handled by putting a wrapper type
+// around the PathValidator, we use the `PathValidatorTabooWrapper` for
+// this. Therefore, the implementation can be rather compact here and
+// uses lots of other functionality to do the heavy lifting.
+// Note that some effort is necessary for memory management. Since we
+// repeatedly call ShortestPath we must constantly reset our subobject,
+// which means that we run into the problem that old references to vertex
+// and edge IDs used in previous path results become invalid. Therefore,
+// we have to copy all these IDs to our own managed memory area.
+// This emplate is instantiated in 8 different ways across the following
+// dimensions:
+//   SingleServerProvider/ClusterProvider
+//   Unweighted/weighted
+//   non-tracing/tracing
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
```

Comment:
Again, I simply used the same approach here as `TwoSidedEnumerator` and  `WeightedTwoSidedEnumerator` do. I am also not sure how this could sensibly be done with actual compiler assistence (maybe with concepts?).

However, I completely sympathize with your view, since I was in exactly the same situation a few weeks ago not understanding anything about these complex template classes here. And I would also have wished for some more information in the source code.

Let's do the following, I add source code comments with explanations about the possible template parameters.

Reply:
Actually, these source code comments are already in place. What else is needed?